### ISSUE TITLE
Linear regression - Random forest

### ISSUE THREAD

**Issue by priyanga24** [awslabs/autogluon#500]
Hi,
why do I get a messy scatterplot while using random forest regressor for simple linear regression with two variables?
![image](https://user-images.githubusercontent.com/34644984/83882161-48f0b180-a774-11ea-9ac1-6312eb6928ad.png)

Code:
from autogluon.utils.tabular.ml.trainer.auto_trainer import AutoTrainer
from autogluon.utils.tabular.ml.models.lr.lr_model import LinearModel
import autogluon as ag
import matplotlib.pyplot as plt
from autogluon import TabularPrediction as task
hyperparameters = {   
           'RF': {} 
    }

train_data2 = task.Dataset(file_path='no_train.csv')
label_column = 'salary'
predictor = task.fit(
        train_data=train_data2,
        label=label_column,
        hyperparameters=hyperparameters)
test_data2 = task.Dataset(file_path='no_test.csv')
y_test2 = test_data2[label_column]  # values to predict
test_data_nolab2 = test_data2.drop(labels=[label_column],axis=1) # delete label column
y_pred2 = predictor.predict(test_data_nolab2)

print("Predictions:  ", y_pred2)
print("actual", y_test2)
print("rff",test_data2['experience'])
perf = predictor.evaluate_predictions(y_true=y_test2, y_pred=y_pred2, auxiliary_metrics=True)
predictor.fit_summary()
plt.scatter(test_data2[["experience"]], y_test2, color = 'red')
plt.plot(test_data2[["experience"]], y_pred2, color = 'blue')
plt.title('experience vs salary')
plt.xlabel('experience')
plt.ylabel('salary')
plt.show()

Only if I give 'LR' in hyperparameter, I can see the best fit regression line correctly. When I use RF regressor, catboost regressor or lightgb, regressor, the plot is shabby. Can I know why is that? SO it means there is not plot for other models, we can examine only the RMSE values? Please help, thanks

**Comment by priyanga24** [comment]
The dataset used was just a random linear regression dataset with one dependent variable(salary) and one independent variable(experience). So based on the experience, the LR model calculates the salary.

**Comment by Innixma** [comment]
This appears to be a tiny dataset, which is not well suited to tree algorithms which don't generalize beyond the bounds of their training data. I suspect if you increased the training data size by 100x, you would see the expected pattern. Note that this isn't particular to AutoGluon, but is a general property of tree algorithms.

**Comment by Innixma** [comment]
Also, the tree algorithm must have had a training example near the 0 experience range to correctly predict it, as tree algorithms do not generalizee beyond their training set bounds.

**Comment by priyanga24** [comment]
Thanks a lot!

**Comment by priyanga24** [comment]
I used the code below to use random forest as regressor . As you said, I extended the dataset size to 12,000 rows in train and 3000 rows in train. But Still I am getting a messy plot as shown below. Please help!

![image](https://user-images.githubusercontent.com/34644984/83938751-c9adbd00-a809-11ea-950b-a8a02223b6dc.png)

**CODE:**
import autogluon as ag
import matplotlib.pyplot as plt
from autogluon import TabularPrediction as task
from autogluon.utils.tabular.metrics import roc_auc
from autogluon.utils.tabular.ml.trainer.auto_trainer import AutoTrainer
from autogluon.utils.tabular.ml.models.lr.lr_model import LinearModel
train_data = task.Dataset(file_path=r'15_train.csv')
label_column = 'salary'
hyperparameters = {'RF': {} } 
predictor = task.fit(train_data=train_data, label=label_column,hyperparameters = hyperparameters)
test_data = task.Dataset(file_path=r'15_test.csv')
y_test = test_data[label_column]  # values to predict
test_data_nolab = test_data.drop(labels=[label_column],axis=1) # delete label column to prove we're not cheating
y_pred = predictor.predict(test_data_nolab)
#plt.plot(test_data[["YearsExperience"]], y_pred, color = 'blue')
#print("Predictions:  ", y_pred)
#print("actual", y_test)
perf = predictor.evaluate_predictions(y_true=y_test, y_pred=y_pred, auxiliary_metrics=True)
resultst = predictor.fit_summary()

Used this reference to generate datasets  https://www.gaussianwaves.com/2020/01/generating-simulated-dataset-for-regression-problems-sklearn-make_regression/

**Comment by Innixma** [comment]
I'd recommend you try passing this same data in to a sklearn random forest model directly without using AutoGluon. If it gives you similar results, then it is a fundamental property of the tree model. If you get your ideal solution however, it is a property of AutoGluon and I can try to assist in resolving it.

Another option: Try specifying hyperparameters = {'RF': {'n_estimators': 10000}} or increasing the data size by another 10x. Also measure the average error of the model, it is likely not much worse than linear regression.

**Comment by priyanga24** [comment]
Hey Nick, Yes you were right, the tree algorithms like RF,KNN and boosting algorithms CAtboost and GBM have a non linear nature and they cannot extrapolate - this is clarified.
But neural networks can work very well with linear regression with specifying linear activation function in the output layer. where can i specify that the activation function for the output layer is 'linear'? But i see that Autogluon does not provide linear activation function with tabular NN, nor does it do by itself , why is it? will you be adding the functionality? reference: https://towardsdatascience.com/neural-networks-with-numpy-for-absolute-beginners-part-2-linear-regression-e53c0c7dea3a (Please scroll down to the very last part)

**Comment by Innixma** [comment]
@jwmueller Perhaps you could answer the NN question?

**Comment by jwmueller** [comment]
If you want to train only a linear model, you can call autogluon like this:

```
task.fit(..., hyperparameters={'LR':{}})
```
Note this is recently-added functionality so please upgrade your autogluon version: `pip install autogluon --upgrade`

If you want a more flexible model that can extrapolate a bit, then yes you can use the NN and call autogluon like this:
``` 
import numpy as np
task.fit(..., hyperparameters={'NN':{'y_range':(-np.inf, np.inf)} }
```
Note the autogluon NN already uses linear activation in the output layer for regression, so there's nothing you need to specify to get that.  The reason you should pass in `y_range` if you desire extrapolation, is because autogluon's NN is actually designed not extrapolate much by default (default 'y_range' will be constrained to y-value range in the training data +- a little bit).  If you want to control how much extrapolation happens, then rather than specifying the `y_range` hyperparameter, you can instead specify the `y_range_extend` hyperparameter (default is 0.05, you can increase it as much as you like to get more aggressive extrapolation):
https://github.com/awslabs/autogluon/blob/b97211ebcf1b65f0a92b31b10c6df50da6df9e44/autogluon/utils/tabular/ml/models/tabular_nn/hyperparameters/parameters.py#L50

For example:
``` 
task.fit(..., hyperparameters={'NN':{'y_range_extend': 1.0} }
```

**Comment by priyanga24** [comment]
ok, I tried changing different values for y_range_extend parameter,still my plot does not show a best fit line.
![image](https://user-images.githubusercontent.com/34644984/84175491-a8232e80-aab2-11ea-9bdd-f13fc50fd80f.png)
if I increase the dataset size also, I dont get a best fit line
![image](https://user-images.githubusercontent.com/34644984/84175705-efa9ba80-aab2-11ea-9a1e-afb0fc12cfb6.png)
So, it is recommended that we use linear regression models for solving linear regression problems rather than using boosting,tree,NN algorithms? as NN is not extrapolating  @jwmueller

**Comment by jwmueller** [comment]
Sorry how can I tell if it is extrapolating from these plots? Are the green dots the training data?  If so to see whether the model is extrapolating, you would need to plot 'predicted' for new X that are bigger/smaller than the biggest/smallest X in the training data.

And yes it is recommended to set "LR" in `hyperparameters` argument of `task.fit()` if you have a linear regression problem.
